core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -266,397 +258,396 @@ private[zio] final class FiberContext[E, A](
+                      case t: Throwable if !platform.fatal(t) => ZIO.fail(t.asInstanceOf[E])
+                    }
 
-                case ZIO.Tags.EffectSuspendTotalWith =>
-                  val zio = curZio.asInstanceOf[ZIO.EffectSuspendTotalWith[Any, E, Any]]
+                  case ZIO.Tags.EffectSuspendTotalWith =>
+                    val zio = curZio.asInstanceOf[ZIO.EffectSuspendTotalWith[Any, E, Any]]
 
-                  val k = zio.f
-                  if (traceExec && inTracingRegion) addTrace(k)
+                    val k = zio.f
+                    if (traceExec && inTracingRegion) addTrace(k)
 
-                  curZio = k(platform)
+                    curZio = k(platform)
 
-                case ZIO.Tags.Trace =>
-                  curZio = nextInstr(captureTrace(null))
+                  case ZIO.Tags.Trace =>
+                    curZio = nextInstr(captureTrace(null))
 
-                case ZIO.Tags.FiberRefNew =>
-                  val zio = curZio.asInstanceOf[ZIO.FiberRefNew[Any]]
+                  case ZIO.Tags.FiberRefNew =>
+                    val zio = curZio.asInstanceOf[ZIO.FiberRefNew[Any]]
 
-                  val fiberRef = new FiberRef[Any](zio.initialValue, zio.combine)
-                  fiberRefLocals.put(fiberRef, zio.initialValue)
+                    val fiberRef = new FiberRef[Any](zio.initialValue, zio.combine)
+                    fiberRefLocals.put(fiberRef, zio.initialValue)
 
-                  curZio = nextInstr(fiberRef)
+                    curZio = nextInstr(fiberRef)
 
-                case ZIO.Tags.FiberRefModify =>
-                  val zio = curZio.asInstanceOf[ZIO.FiberRefModify[Any, Any]]
+                  case ZIO.Tags.FiberRefModify =>
+                    val zio = curZio.asInstanceOf[ZIO.FiberRefModify[Any, Any]]
 
-                  val oldValue           = Option(fiberRefLocals.get(zio.fiberRef))
-                  val (result, newValue) = zio.f(oldValue.getOrElse(zio.fiberRef.initial))
-                  fiberRefLocals.put(zio.fiberRef, newValue)
+                    val oldValue           = Option(fiberRefLocals.get(zio.fiberRef))
+                    val (result, newValue) = zio.f(oldValue.getOrElse(zio.fiberRef.initial))
+                    fiberRefLocals.put(zio.fiberRef, newValue)
 
-                  curZio = nextInstr(result)
+                    curZio = nextInstr(result)
 
-                case ZIO.Tags.RaceWith =>
-                  val zio = curZio.asInstanceOf[ZIO.RaceWith[Any, Any, Any, Any, Any, Any, Any]]
-                  curZio = raceWithImpl(zio).asInstanceOf[IO[E, Any]]
+                  case ZIO.Tags.RaceWith =>
+                    val zio = curZio.asInstanceOf[ZIO.RaceWith[Any, Any, Any, Any, Any, Any, Any]]
+                    curZio = raceWithImpl(zio).asInstanceOf[IO[E, Any]]
+                }
               }
+            } else {
+              // Fiber was interrupted
+              curZio = ZIO.halt(state.get.interrupted)
```

Comment:
Ok, if `.refailWithTrace`, why not `.haltWith`? As is there's no info on what the fiber has been executing immediately before interruption

Reply:
It can't be a composite instruction or it will end in an infinite loop. So we have to build the full cause right here. I could attach the trace here, I suppose?

Reply:
@jdegoes Yeah you can construct with one op:

```scala
val snapshot = state.get.interrupted
ZIO.haltWith(trace => Cause.traced(snapshot, trace())
```

Reply:
Actually I see `ZIO.halt` will generate `Fail` which will be filled with the trace. Should be good enough, right?

Reply:
@jdegoes `fail` and `die` do generate the trace, but`halt` passes the raw Cause, see impl - https://github.com/zio/zio/blob/4b7225dd36f6a4b16f170a778cb553e18cb461d1/core/shared/src/main/scala/zio/ZIO.scala#L2279